lib/http2/http2_debug_state.c

```diff
@@ -0,0 +1,212 @@
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include <inttypes.h>
+#include "h2o.h"
+#include "h2o/http2.h"
+#include "h2o/http2_internal.h"
+
+static const char *debug_state_string_open = "OPEN";
+static const char *debug_state_string_half_closed_remote = "HALF_CLOSED_REMOTE";
+static const char *debug_state_string_reserved_local = "RESERVED_LOCAL";
+
+static const char *get_debug_state_string(h2o_http2_stream_t *stream)
+{
+    if (h2o_http2_stream_is_push(stream->stream_id)) {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+            return debug_state_string_reserved_local;
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    } else {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+            return debug_state_string_open;
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+__attribute__((format(printf, 3, 4))) static void append_chunk(h2o_mem_pool_t *pool, h2o_iovec_vector_t *chunks, const char *fmt, ...)
+
+{
+    va_list args;
+
+    va_start(args, fmt);
+    int size = vsnprintf(NULL, 0, fmt, args);
+    va_end(args);
+
+    assert(size > 0);
+
+    h2o_iovec_t v;
+    v.base = h2o_mem_alloc_pool(pool, size + 1);
+
+    va_start(args, fmt);
+    v.len = vsnprintf(v.base, size + 1, fmt, args);
+    va_end(args);
+
+    h2o_vector_reserve(pool, chunks, chunks->size + 1);
+    chunks->entries[chunks->size++] = v;
+}
+
+static void append_header_table_chunks(h2o_mem_pool_t *pool, h2o_iovec_vector_t *chunks, h2o_hpack_header_table_t *header_table)
+{
+    int i, comma_removed = 0;
+    for (i = 0; i < header_table->num_entries; i++) {
+        h2o_hpack_header_table_entry_t *entry = h2o_hpack_header_table_get(header_table, i);
+        append_chunk(pool, chunks,
+                    ",\n"
+                    "      [ \"%.*s\", \"%.*s\" ]",
+                    (int)entry->name->len, entry->name->base,
+                    (int)entry->value->len, entry->value->base);
+        if (!comma_removed) {
+            chunks->entries[chunks->size - 1].base[0] = ' ';
```

Comment:
It might be preferable to do the following instead so as not to leave trailing spaces.
```
chunks->entries[chunk->size - 1].base += 1;
chunk->entries[chunk->size - 1].len -= 1;
```

It _might_ be better to define the logic as a separate function, since you would want to do the same at other parts of the code as well.